오픈 마켓 [STEP 2] 써니쿠키, 메네

sunny-maeng opened the pull request:
@inwoodev
안녕하세요 제임스! 🙇🏻‍♀️🙇
Step2 완료하고 PR드립니다.
Modern Collection Views에 대해 공부하고 사용해보느라 PR이 많이 늦은 점 죄송합니다.
이번에도 잘 부탁드립니다. 많이 알려주세요! 🙏🏻

## 💭 고민한 부분
### 1️⃣ 썸네일 이미지의 비동기 작업
서버에서 가져오는 썸네일 이미지의 url값을 이용해 사진을 넣어줄 때, `init(contentsOf:)`메서드를 사용해 Data를 가져왔는데 공식문서에 네트워크 기반 URL을 이용해 이 메서드를 사용 할 때는 비동기 작업으로 처리해주라하여 URLSession을 이용한 `fetchImage`메서드를 만들어 비동기로 처리했습니다.

<br>

### 2️⃣ 빠르게 스크롤 했을 대 Cell의 위치가 뒤죽박죽 섞이는 문제가 있었습니다.
사진을 비동기로 가져오게 되면서, List와 Grid에서 스크롤을 빠르게 했을 때, Cell의 위치가 뒤죽박죽되고 다른 이미지가 들어있다가 순식간에 다시 정상이미지가 돌아오는 등, Cell이 제자리를 찾지 못하는 문제가 있었습니다.</br>

<img width = 150, src = "https://i.imgur.com/xe6R3pE.gif"> </br>

cell 이 제자리를 찾을 수 있도록 indexPath를 이용했습니다. UICollectionView.CellRegistration`메서드에서 사용할 수 있는 `indexPath값과 `UICollectionView`의 indexPath 값이 같을 때에만 cell의 content를 생성하도록 하였습니다.

처음엔 이 비교를 위해 셀의 컨텐츠를 구성하는 기능을 하는 `configureCell` 메서드에서 3가지 `UICollectionView`, `Cell타입`, `IndexPath` 를 추가로 매개변수로 받아야 했습니다..
매개변수사용 대신 `completionHandler` 로 `@escaping` 클로저를 사용해 뷰컨에서 indexPath를 비교하도록 수정과정을 거쳤습니다.

```swift
// class MarketListCell
 func configureCell(page: Page,
                       completionHandler: @escaping (() -> Void) -> Void) {
     
     //  ... (셀의 콘텐츠를 구성)

    DispatchQueue.main.async {
        let updateConfiguration = {
            self.pageListContentView.configuration = content
    }
        completionHandler(updateConfiguration)
    }
 }

// class MarketListViewController
private func configureDataSource() {
    let cellRegistration = UICollectionView.CellRegistration<MarketListCell, Page> {
        (cell, indexPath, page) in
        cell.configureCell(page: page) { updateConfiguration in
            if indexPath == self.listView.indexPath(for: cell) {
                updateConfiguration()
            }
        }
    //  ... 
```

<br>

### 3️⃣ 기본 이미지가 등록되어 있지 않아 로딩 시 셀의 크기가 각각 다른 문제
Modern Collection Views에서는 셀의 사이즈가 자동으로 지정되어 List뷰 로딩 시 레이블의 크기만큼 셀의 크기가 먼저 정해지고 이미지가 로딩되면 이후에 셀이 커지면서 각각 다른 크기의 셀로 작동되는 문제가 있었습니다.

<img width = 250, src = "https://i.imgur.com/oN1iL7v.png"></br>


`content.imageProperties.reservedLayoutSize` 프로퍼티를 이용하여 사이즈를 지정하고, 
로딩 이미지를 `PlaceHolder`로 넣어주어 셀의 크기가 먼저 잡히도록 처리해 주었습니다.
```swift
content.image = UIImage(named: "loading")
content.imageProperties.reservedLayoutSize = CGSize(width: 70, height: 70)
content.imageProperties.maximumSize = CGSize(width: 70, height: 70)
```

<br>

### 4️⃣ 요구사항 List Cell 우측 "잔여수량: " 우측의 `>`를 위쪽으로 배치하기 위해 고민해 보았습니다.

|요구사항|Disclosure<br>이용 시|
|:--:|:--:
|![](https://i.imgur.com/WbKY4EI.png)|![](https://i.imgur.com/XnSpod7.png)|

잔여수량 label과 우측 > label을 `StackView`로 담아주고 Alignment를 `Top`으로 설정해 주었습니다.

### 5️⃣ 레이블에 ~~취소선~~과 색상을 주기 위해 고민해 보았습니다.
`AttributedString`을 사용하여 취소선과 할인 가격, 품절 색상을 적용하여 주었는데 `UILabel`을 extention하여 label에 적용하는 방법과 `NSMutableAttributedString`을 extention하는 방법 중 어떤 것을 사용할지 고민했습니다.
  ➡️ ListView에서 label에 직접 접근할 수 없어서 `NSMutableAttributedString`을 익스텐션하여 적용해주고 싶은 설정을 `append`로 연결하여 출력해 주었습니다.

### 6️⃣ 이미지다운의 비동기작업으로 인한 딜레이의 화면표현방법과 캐싱작업
이미지 마다 `UIActivityIndicatorView`를 이용해 이미지 로딩을 표현하고 싶었는데, `UICollectionViewListCell의` 기본 셀인 `.subtitleCell`을 사용하는 구성이라 이용에 어려움이 있었습니다. 
`UIListContentConfiguration`을 이용해 특성을 잡아줘야하는데 이미지뷰대신 indicator를 넣어주려니 타입문제와, addSubView가 되지않는 문제가 있어 적용하지 못했습니다.

`LaunchScreen`도 생각해 보았습니다. 처음에 런치스크린을 노출해 주고 데이터를 받아오는 시간을 벌어보려 했는데 HIG 문서에서 보여줄 수 있는 정보는 먼저 보여주라고 쓰여있는 것을 보았던 기억이 나서 적용을 보류했습니다.

최종적으론, 로딩 image하나를 Asset으로 등록하여 작업전까지 로딩이미지가 보이도록 처리했습니다
더불어 작업 성능을 위해 이미지를 캐싱해놓도록 했습니다.

---

## 🧐 조언을 얻고 싶은 부분

### ❓ 1️⃣ cell의 콘텐츠를 구성해주는 메서드의 양이 많아서 기능분리를 해야하는지 고민됩니다
List와 Grid의 각 Cell의 내용을 구성해주는 `configureCell`은 cell에 이미지, 이름, 가격을 담아주는 역할을 합니다.
내용 구성을 하다보니 이미지의 캐싱작업, 이미지의 비동기작업, 이미지의 프레임작업, 이미지의 placeHolder넣어주기 같은 작업들도 필요하다보니 길이가 길어서 기능분리를 해줘야하는 것인지.. 하나의 기능으로 생각하여 이대로 내용이 길지만 묶어서 가지고있어도 될 지 모르겠습니다.. 
특히 `ListCell`에서는 `Configuration`을 이용해 특성들을 캡슐화하는것이니 괜찮다는 생각도 드는데 코드 길이가 길다보니 어느 방향이 더 좋을지 모르겠습니다.. 제임스라면 어떻게 하셨을지 궁금합니다 조언 부탁드리겠습니다! 

<br>

### ❓ 2️⃣ 제목이 길어지는 경우 어떻게 보여주는 편이 좋을지 조언을 듣고 싶습니다.

Grid뷰에서 제목이 긴 경우, 현재는 말줄임 표시를 해서 보여주고 있는데 `numberOfLines`를 0으로 변경하면 줄바꿈하여 전체를 보여줄 수 있지만 셀의 크기를 잡아주었기 때문에 아래쪽 Label이 잘려서 보여지게 됩니다.
이 문제를 해결하기 위해 셀의 크기를 크게 준다면 아래쪽으로 비어있는 공간이 많이 노출될 것 같습니다.

|말줄임 X|말줄임 O|
|:--:|:--:|
|![](https://i.imgur.com/NYU8AiL.png)|![](https://i.imgur.com/7VFlO2C.png)|

이런 것은 기획의 문제이지 않을까 싶은데요,
등록 화면에서 (아직 구현이 되어있진 않지만)글자수 제한을 둔다거나 하는 방법으로 어느정도 해결할 수 있을 것 같은데 제임스의 생각이 궁금합니다.

JaeKimdev committed:
design: 뷰 이동을 위한 segmentControl과 신규 등록 버튼 스토리보드에서 생성

sunny-maeng committed:
feat: ModernCollectionView의 리스트형구현을 위한 Configuration Key등록 및 State프로퍼티 등록

JaeKimdev committed:
feat: MarketCollectionViewListCell 셀 contentView 구현 및 레이아웃 지정

sunny-maeng committed:
feat: MarketCollectionViewListCell의  updateConfiguration 메서드를 override하여 Cell의 컨텐츠구성

JaeKimdev committed:
feat: MarketListViewController에서 레이아웃 생성과 적용하도록 구현

- createListLayout() : UICollectionViewCompositionalLayout 생성해주는 메서드
- configureCollectionView : 생성한 레이아웃을 뷰에 추가해주는 메서드

sunny-maeng committed:
feat: MarketListViewController에 DataSource생성, configureCollectionView메서드내부에서 Cell data전달

JaeKimdev committed:
feat: MarketListViewController 호출부 구현 및 item 세부사항 수정

sunny-maeng committed:
feat: MarketGridViewController에서 컬렉션뷰 layout잡는 메서드 구현
createGridLayout() ->  UICollectionViewCompositionalLayout

JaeKimdev committed:
feat: MarketGridVC에서 컬렉션 뷰를 생성하고  레이아웃 잡는 configureCollectionView 메서드 구현

sunny-maeng committed:
feat: MarketCollectionViewGridCell에서 이미지, 레이블 생성

JaeKimdev committed:
feat: MarketCollectionViewGridCell 구성요소 오토레이아웃 설정

sunny-maeng committed:
feat: MarketCollectionViewGridCell에서 Cell에 page정보를 담아줄 configureCell()메서드 구현

JaeKimdev committed:
feat: MarketGridViewController타입의 configureDataSource()메서드 구현

sunny-maeng committed:
feat: MarketGridViewController에서 서버에서 데이터를 가져오는 fetchMarketData 메서드 구현

JaeKimdev committed:
feat: MarketCollectionViewGridCell 내부에 StackView 추가 및 레이아웃, 폰트 수정

sunny-maeng committed:
feat: NSMutableAttributedString extension에 할인하는 상품의 할인가를 표시하고 정가에 취소선을 긋는 메서드 구현

JaeKimdev committed:
design: MarketCollectionViewGridCell 아이템 spacing 및 borderline 수정

sunny-maeng committed:
feat: 품절 표시에 노란색으로 표기하는 yellowColor()메서드 구현 및  AttributedString 만드는 메서드들 네이밍 수정

JaeKimdev committed:
design: segmentControl 선택시 색상변경 및 ListCell 레이블 레이아웃 수정

sunny-maeng committed:
design: 셀 악세사리로 사용하던 disclosureIndicator를 View로 만든 후 위치 조정

JaeKimdev committed:
design: MarketCollectionViewListCell 구성요소 레이아웃 수정

sunny-maeng committed:
feat: 이미지 다운로드 비동기처리하는 fetchImage() 메서드 구현 및 적용

JaeKimdev committed:
feat: 셀 높이 고정을 위해 loading 이미지 placeHolder로 추가

sunny-maeng committed:
feat: MarketCollectionViewListCell에서 state기능 제거 및 cell이 알맞는 indexPath에 생성되는 로직 추가

JaeKimdev committed:
feat: 이미지 캐싱을 위한 ImageCacheProvider 싱글톤 생성 후, fetchImage 메서드에 적용

sunny-maeng committed:
feat: MarketCollectionViewGridCell에 cell이 알맞는 indexPath 에 생성되는 로직 추가

JaeKimdev committed:
refactor: 파일 컨벤션 수정 및 메서드 네이밍 수정, 접근제어 설정

sunny-maeng committed:
refactor: configureCell메서드에서 indexPath비교를위한 매개변수들을 completionHandler 사용으로 수정

sunny-maeng committed:
Update README.md

inwoodev commented on OpenMarket/OpenMarket/Model/URLSession/URLSessionProvider.swift:
```diff
@@ -6,6 +6,7 @@
 //
 
 import Foundation
+import UIKit
```
Foundation과 UIKit 두 개 다 import할 필요가 있을까요? :)

JaeKimdev replied:
앗! 여기서는 미처 확인을 못했네요...😭
리팩토링 시 확인후 제거하겠습니다.

JaeKimdev replied:
https://github.com/yagom-academy/ios-open-market/pull/221/commits/caf6bde9a5c012772885c90b3f185755f2e0d4ed
오토레이아웃 제약 제거하면서 함께 수정하였습니다.

inwoodev commented on OpenMarket/OpenMarket/ViewController/Extension/NSMutableAttributedString+Extension.swift:
```diff
@@ -0,0 +1,37 @@
+                                      range: NSMakeRange(0, attributedString.length))
+        
+        return attributedString
+    }
```
👍👍👍

inwoodev commented on OpenMarket/OpenMarket/Model/DataType/Page.swift:
```diff
@@ -7,20 +7,20 @@
     }
 
-    struct Vendors: Decodable {
+    struct Vendors: Decodable, Hashable {
```
DiffableDataSource에서 사용하기 위해서는 `Hashable`채택이 필수이죠 ㅎㅎ. DiffableDataSource에 대해서 얼마나 공부하셨는지 궁금해요~ 습득하신 지식을 저에게도 알려주시면 감사하겠습니다~

JaeKimdev replied:
DiffableDataSource는 CollectionView에 셀을 제공하고 데이터를 관리하기 위한 객체입니다
`SectionIdentifierType`, `ItemIdentifierType`을 가지는 Snapshot을 가지고 변경사항을 비교하는데 이 두개의 파라미터의 hash value를 가지고 비교하기 때문에 `Hashable`해야 합니다.
데이터 자동  동기화로 에러를 방지할 수 있고, 데이터가 변경될 때 자동으로 변경 애니메이션을 적용하여 줍니다.

sunny-maeng replied:
메네의 의견과 같습니다... 기존 DataSource와 다르게 DiffableDataSource는 **hashValue**를 통해 항목 업데이트 때마다 업데이트된 컬렉션과 이전에 표시된 컬렉션 간의 차이를 자동으로 계산하고, 새로운 부분만 다시 그릴 수 있습니다. 
 이 때, 이전 DataSource 를 사용 할 때엔 `reloadData()` 메소드를 사용했어서 애니메이션 효과를 주기 힘들었지만, DiffableDataSource는 애니메이션 효과로(= 공식문서에서의 표현은 '자연스럽게'라고 표현) 사용자 경험을 높일 수 있다는 장점도 가지고있습니다.

inwoodev commented on OpenMarket/OpenMarket/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,40 @@
+                                              for: .normal)
+    }
+    
+    @IBAction private func changeView(_ sender: UISegmentedControl) {
```
changeView보다 더 나은 메서드 네이밍은 분명히 존재할거같아요!!😉

sunny-maeng replied:
switchLayout으로 수정했습니다 :) !
(https://github.com/yagom-academy/ios-open-market/pull/221/commits/1ce6f07e25143570d2b16aa7788b8ee1638b43f6)

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketGridCell.swift:
```diff
@@ -0,0 +1,131 @@
+                }
+            }
+        }
+    }
```
우려하시는데로 해당 메서드는 너무 많은 작업을 담당하는 것 같아요.
해당 메서드는:
- 상품 가격에 따라 label에 다른 텍스트를 지정하며
- 상품 수량에 따라 label에 다른 텍스트를 지정하고
- 캐시이미지 체크 후 유무에 따라 캐싱 또는 네트워킹 작업까지
메서드 분리가 필요 해 보이는데 어떻게 생각하시나요~?

JaeKimdev replied:
4b0a124 에서
Price, Stock 레이블 작업을 하는 메서드와 사진 작업을 하는 메서드로 각각 분리 하였습니다.
005bef9 에서
fetchImage, generatePriceLabelContent, generateStockLabelContent ListCell에 맞추어 메서드 네이밍 변경하였습니다.

inwoodev replied:
조금 더 가독성이 좋아졌네요~ 👍

inwoodev commented on OpenMarket/OpenMarket/Model/URLSession/URLSessionProvider.swift:
```diff
@@ -47,4 +48,31 @@ final class MarketURLSessionProvider {
+        }
+        
+        dataTask.resume()
+    }
```
만약 서버로부터 오디오데이터를 불러오는 작업이 필요하다 하면 `MarketURLSessionProvider`가 또 사용되겠죠? 그리고 오디오데이터를 가져와야 함에도 우리는 `fetchImage()`메서드를 사용할 수 있게 됩니다. 그럼 결론적으로 위 객체는 범용성이 낮은 객체가 된다고 저는 생각합니다. 오디오를 가져오는 객체에서 위 객체를 사용하게 되면 결과적으로 solid원칙을 위배하기 때문이죠. 또한 오디오를 가져오는 메서드를 추가적으로 하나 더 만들어야 하니 결국 MarketURLSessionProvider는 짬뽕 객체가 될 가능성이 높아보여요.

저같으면 MarketURLSessionProvider를 활용하여 이미지를 가져오는 객체를 별도로 만들거 같아요. 두 분의 의견도 궁금합니다~

JaeKimdev replied:
371605d
fetchData 메서드의 범용성을 높여서 mimeType 확인하는 부분을 삭제하고 data 타입으로 반환하도록 수정하였습니다.
받아온 data를 가공하여 여러가지로 활용할 수 있도록 수정하였습니다.
이렇게 하면 단일책임 원칙도 지킬 수 있을것 같습니다. 😄

inwoodev commented on OpenMarket/OpenMarket/ViewController/MainViewController.swift:
```diff
@@ -0,0 +1,40 @@
+            gridView.isHidden = false
+        }
+    }
+}
```
enum을 적절히 활용 해 볼 수 있지 않을까요?

```suggestion
enum SegmentedItemState {
case grid
case list
}
.
.
.
guard let selectedIndex = SegmentedItemState(rawValue: sender.selectedSegmentIndex) else { return }

switch selectedIndex {
// hidden 처리
}
```

sunny-maeng replied:
조언주신데로 SegmentedItemState 를 enum으로 정리했습니다 ㅎㅎ
enum으로 정리해놓으면 유지보수에 더 좋을 것 같고, switch문에서 list, grid 라고 쓸 수 있으니 명시적으로 느껴져서 너무 좋네요! 감사합니다 :)
 (https://github.com/yagom-academy/ios-open-market/pull/221/commits/1ce6f07e25143570d2b16aa7788b8ee1638b43f6)

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketGridCell.swift:
```diff
@@ -0,0 +1,131 @@
+        stackView.axis = .vertical
+        stackView.alignment = .center
+        stackView.spacing = 10
+        stackView.translatesAutoresizingMaskIntoConstraints = false
```
`translatesAutoresizingMaskIntoConstraints`에 대해서 잘 알고 계신지 궁금합니다. stackview에서는 필요한데 label에서는 왜 설정하지 않으셨나요~?

JaeKimdev replied:
코드로 뷰를 그려본 적이 거의 없어서 false로 설정해주는 것을 안하고 넘어간 것 같습니다. ㅠㅠ
리팩토링 시 수정하겠습니다.

sunny-maeng replied:
알고있는것과 공부한것을 적고 질문을 남기겠습니다.. 

`translatesAutoresizingMaskIntoConstraints`의 기능은 이름에도 적혀있듯이`AutoresizingMask`와 관련된 속성입니다.
AutoresizingMask란, View안에 View가 있을 때 `superview`의 bounds 크기가 변하면 `subview`가 자신의 크기나 위치를 조정하는 방식을 결정해야 하는데, 이 때 쓰이는 개념으로 자신의 크기를 조정하는 방법을 결정하는 비트마스크입니다. 

**오토레이아웃과 비슷한 기능이기 떄문에 오토레이아웃과 오토리사이징은 충돌**할 수 있습니다.  

만약 코드로` Autolauout 제약`을 걸어주게되면  `autoresizingMask`가 하는 기능과 충돌하는 것이고, 이를 방지하기 위해 `Auturesizing`을 사용하지 않는다고 명시해주는게 `translatesAutoresizingMaskIntoConstraints = false`라고 알고 있습니다.

---

사실 처음에는 라벨과 이미지 모두에 `translatesAutoresizingMaskIntoConstraints = false` 를 지정해줬다가, 테스트겸 어느부분에서 꼭 필요한지 코드를 한군데씩 지워보고 시뮬레이션을 돌려보았었습니다..  결과적으론 직접 오토레이아웃을 걸어주는 `stackView`는 `translatesAutoresizingMaskIntoConstraints`속성을 `false`로 줘야 정상작동하지만, `stackView`안에 담기는 `SubView`들은 `translatesAutoresizingMaskIntoConstraints` 속성을 따로 지정해주지 않아도 정상작동이고 에러문구도  나오지않아서 코드를 삭제했었습니다.

`stackView의 subView들`은  stackView의 `distribution` , `alignment` 에 강한 영향을 받아서 레이아웃이 정렬 되는건가?? 하고 어림짐작중이었습니다....  이런경우라도 모든 (스택의 서브뷰 포함) view에 `translatesAutoresizingMaskIntoConstraints = false` 지정을 해주는게 더 안전할거라 생각하고 지정해줘야하는 걸까요,,!?

inwoodev replied:
제가 공부하기로는 모든 뷰에 대해 해당 프로퍼티는 true로 기본적으로 설정되는 것으로 알고 있습니다. 그렇기 때문에 별도의 프레임 레이아웃 지정을 하지 않아도 화면에 얼추 띄울 수 있는 것이지요. 하지만 자동으로 리사이징을 해주는 해당 프로퍼티와 프레임 기반으로 계산된 크기와 위치간의 충돌을 방지하기 위해서라도 저 같은 경우 프레임 기반으로 제약을 줄 경우 false 설정을 필수로 해야 한다고 생각해요 :) 스토리보드는 왜 이 프로퍼티를 false로 하는 작업이 필수가 아닌지도 혹시 아시나요 ㅎㅎ

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketListCell.swift:
```diff
@@ -0,0 +1,127 @@
+    private var stockStackView: UIStackView = {
+        let stackView = UIStackView()
+        stackView.distribution = .fillProportionally
+        stackView.spacing = 10
```
여기는 왜 `.translatesAutoresizingMaskIntoConstraints`가 빠져 있을까요~~? `.translatesAutoresizingMaskIntoConstraint` 에 대해 아시는 부분을 공유 해 주시면 감사하겠습니다~~

JaeKimdev replied:
코드로 뷰를 그릴때는`translatesAutoresizingMaskIntoConstraints`를 false로 설정해 주어야 한다. 정도만 알고 있었는데 이번 기회에 공부해 보았습니다.
`translatesAutoresizingMaskIntoConstraints`는 UIView의 인스턴스 프로퍼티로 View의 autoresizing  mask가  Auto Layout constraints으로 변환되는지 설정하는 Bool 타입 값입니다.

autoresizing mask constraint는 View의 크기와 위치를 완전히 지정하기 때문에 false로 변경하지 않으면 제약을 추가할 수 없습니다.

sunny-maeng replied:
이 부분은, `setupLayout()`메서드 내부에서 addSubview와 함께 forEach문을 돌 때 속성을 주고있습니다.
중복코드를 피하기위해 아래에서만 작성했는데 코드의 통일성을 위해 코드를 옮겨 StacView 속성들과 같이 놓는게 코드 읽는사람들에게 보기가 좋을까요 ??

```swift
private func setupLayout() {
// ...
        [pageListContentView, stockStackView].forEach {
            contentView.addSubview($0)
            $0.translatesAutoresizingMaskIntoConstraints = false
        }
// ...
}
```

inwoodev replied:
취향 차이인거 같아요 :) 이것도 가독성 측면에서는 개인적으로 나쁘지 않은것 같습니다~

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketListCell.swift:
```diff
@@ -0,0 +1,127 @@
+                }
+            }
+        }
+        pageListContentView.configuration = content
```
여기도 기존 configureCell메서드에서 남겼던 코멘트와 비슷한 생각입니다. 메서드를 분리하면 좋을 것 같아요~!

JaeKimdev replied:
005bef9
fetchImage, generatePriceLabelContent, generateStockLabelContent 메서드 분리하였습니다.

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketListCell.swift:
```diff
@@ -0,0 +1,127 @@
+        let cacheKey = NSString(string: thumbnailUrl)
+        let session = MarketURLSessionProvider()
+        
+        content.image = UIImage(named: "loading")
+        content.imageProperties.reservedLayoutSize = CGSize(width: 70, height: 70)
+        content.imageProperties.maximumSize = CGSize(width: 70, height: 70)
+        content.imageProperties.cornerRadius = 10
```
이미지 문제를 훌륭하게 해결하신 부분이군요 👍

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketGridViewController.swift:
```diff
@@ -0,0 +1,105 @@
+        }
+    }
+    
+    private func setupGridLayout() -> UICollectionViewCompositionalLayout {
+        let itemSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1),
+                                              heightDimension: .fractionalHeight(1))
+        let item = NSCollectionLayoutItem(layoutSize: itemSize)
+        
+        item.contentInsets = NSDirectionalEdgeInsets(top: 10, leading: 10, bottom: 10, trailing: 10)
+        
+        let groupSize = NSCollectionLayoutSize(widthDimension: .fractionalWidth(1),
+                                               heightDimension: .estimated(350))
+        let group = NSCollectionLayoutGroup.horizontal(layoutSize: groupSize,
+                                                       subitem: item,
+                                                       count: 2)
+        
+        let section = NSCollectionLayoutSection(group: group)
+        
+        section.contentInsets = NSDirectionalEdgeInsets(top: 0, leading: 10, bottom: 0, trailing: 10)
+        
+        let layout = UICollectionViewCompositionalLayout(section: section)
+        
+        return layout
+    }
+    
+    private func setupGridFrameLayout() {
```
두 메서드의 네이밍이 매우 유사해서 가독성이 조금 떨어지는 것 같아요. 저 같은 경우 메서드가 특정 타입을 반환하는 경우 create, generate과 같은 동사를 앞에 붙이기도 합니다 :)

sunny-maeng replied:
조언주신데로 두 메서드 중 반환값이 있는 메서드에 `generate` 를 붙인 네이밍을 사용했습니다
꿀팁 감사합니다 :) 
(https://github.com/yagom-academy/ios-open-market/pull/221/commits/1ce6f07e25143570d2b16aa7788b8ee1638b43f6)

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketGridViewController.swift:
```diff
@@ -0,0 +1,105 @@
+        snapshot.appendItems(pageData)
+        dataSource?.apply(snapshot)
+    }
+}
```
DataSource 설정과 스냅샷 적용이 하나의 메서드에 있네요. 분리하는 건 어떻게 생각하시나요?

sunny-maeng replied:
데이터소스 설정과 스냅샷적용 메서드 분리했습니다 :) 
더불어 typeAlias 적용했습니다 :)
(https://github.com/yagom-academy/ios-open-market/pull/221/commits/91928b57fac112413c281dd70370a84f5d3b00e5)

inwoodev commented on OpenMarket/OpenMarket/ViewController/MarketListViewController.swift:
```diff
@@ -0,0 +1,80 @@
+        snapshot.appendItems(pageData)
+        dataSource?.apply(snapshot)
+    }
+}
```
비슷한 내용입니다. 기능을 분리하는건 어떤가요:)

sunny-maeng replied:
기능분리 및 typeAlias적용내용 바로 위 리댓내용과 같습니다 !
(https://github.com/yagom-academy/ios-open-market/pull/221/commits/91928b57fac112413c281dd70370a84f5d3b00e5)

inwoodev reviewed (commented):
어려운 구간을 훌륭히 지나고 계신 여러분을 응원합니다. 이번 스텝도 훌륭하게 진행하셨네요. 고민하신 흔적을 확인 해 보니 이제 거의 문제 해결의 달인인거 같습니다!! 👍. 덕분에 저도 많이 배웠습니다. 저도 제가 아는 선에서 최대한 자세하게 답변 남기도록 할게요. 혹시나 이해가 안되는 부분이 있으시다면 언제든 Dm주세요~~

## 조언을 얻고 싶으신 부분에 대한 제 생각:

> ❓ 1️⃣ cell의 콘텐츠를 구성해주는 메서드의 양이 많아서 기능분리를 해야하는지 고민됩니다
> List와 Grid의 각 Cell의 내용을 구성해주는 configureCell은 cell에 이미지, 이름, 가격을 담아주는 역할을 합니다.
내용 구성을 하다보니 이미지의 캐싱작업, 이미지의 비동기작업, 이미지의 프레임작업, 이미지의 placeHolder넣어주기 같은 작업들도 필요하다보니 길이가 길어서 기능분리를 해줘야하는 것인지.. 하나의 기능으로 생각하여 이대로 내용이 길지만 묶어서 가지고있어도 될 지 모르겠습니다..
특히 ListCell에서는 Configuration을 이용해 특성들을 캡슐화하는것이니 괜찮다는 생각도 드는데 코드 길이가 길다보니 어느 방향이 더 좋을지 모르겠습니다.. 제임스라면 어떻게 하셨을지 궁금합니다 조언 부탁드리겠습니다!

코멘트에도 남긴 부분이에요. configureCell의 메서드를 분리할 필요가 있어 보입니다. 기능 별로 메서드를 분리하여 configureCell에서 호출하는 것도 좋은 방법인거 같구요. 또한 별도의 객체로 로직 부분을 래핑하게 해서 cell이 전달받은 뷰만 보여주는 방향도 좀 더 확장 가능한 방향 같아 보입니다.

---

> ❓ 2️⃣ 제목이 길어지는 경우 어떻게 보여주는 편이 좋을지 조언을 듣고 싶습니다.
Grid뷰에서 제목이 긴 경우, 현재는 말줄임 표시를 해서 보여주고 있는데 numberOfLines를 0으로 변경하면 줄바꿈하여 전체를 보여줄 수 있지만 셀의 크기를 잡아주었기 때문에 아래쪽 Label이 잘려서 보여지게 됩니다.
이 문제를 해결하기 위해 셀의 크기를 크게 준다면 아래쪽으로 비어있는 공간이 많이 노출될 것 같습니다.
이런 것은 기획의 문제이지 않을까 싶은데요,
등록 화면에서 (아직 구현이 되어있진 않지만)글자수 제한을 둔다거나 하는 방법으로 어느정도 해결할 수 있을 것 같은데 제임스의 생각이 궁금합니다.

저도 두분의 생각과 비슷하고 이 부분은 기획과 소통해서 해결 해 나가야 할 문제 같아요~ 아니면 정말로 답이 없기 때문이죠 ㅎㅎ. UX적인측면에서 제 생각을 여쭤보신거라면 저는 라벨이 짤리는게 맞다고 생각합니다. 그 이유는 각 수량의 자세한 정보를 파악할 수 있는 페이지로 유저가 이동할 수 있기 때문이죠. 그래서 리스트 같은 경우 이미지를 통해서 초기 정보를 습득하고 만약 해당 상품에 더 관심이 있다면 해당 셀을 클릭해서 자세한 정보를 파악할 수 있는게 조금 더 ux적으로 자연스럽지 않나 생각 해 봤고 이런 측면에서 리스트에서 라벨의 가치가 비교적 떨어진다고 생각하기에 저라면 기획에게 이런 부분을 어필해서 라벨을 자르는건 어떤지 한 번 어필 해 볼거 같아요  :)

---


아 맞다 그리고 앱 실행하다보니 익숙한 문제가 보이더군요 :(
```
(
    "<NSLayoutConstraint:0x6000026a8d70 UIImageView:0x12751b760.width == 0.9*UIView:0x12751e220.width   (active)>",
    "<NSLayoutConstraint:0x6000026a8eb0 H:|-(10)-[UIStackView:0x12751d8f0]   (active, names: '|':UIView:0x12751e220 )>",
    "<NSLayoutConstraint:0x6000026a8f00 UIStackView:0x12751d8f0.trailing == UIView:0x12751e220.trailing - 10   (active)>",
    "<NSLayoutConstraint:0x6000026bc230 'UISV-canvas-connection' H:[_UILayoutSpacer:0x600003aea490'UISV-alignment-spanner']-(0)-|   (active, names: '|':UIStackView:0x12751d8f0 )>",
    "<NSLayoutConstraint:0x6000026bc1e0 'UISV-canvas-connection' UIStackView:0x12751d8f0.centerX == UIImageView:0x12751b760.centerX   (active)>",
    "<NSLayoutConstraint:0x6000026bc500 'UISV-spanning-boundary' _UILayoutSpacer:0x600003aea490'UISV-alignment-spanner'.trailing >= UIImageView:0x12751b760.trailing   (active)>",
    "<NSLayoutConstraint:0x6000026bfd90 'UIView-Encapsulated-Layout-Width' UIView:0x12751e220.width == 166.333   (active)>"
)
```
고질적으로 맞닥뜨리는 문제죠. 오토레이아웃 문제는 해결 하셨으면 합니다~

JaeKimdev committed:
fix: Gird셀 오토레이아웃 중복 제약 제거

sunny-maeng committed:
refactor: SegmentedItemState 타입 구현 및 네이밍수정, gridCell layout 생성하는 메서드 네이밍 수정

JaeKimdev committed:
refactor: URLSessionProvider 메서드 fetchData로 통일하고 받아온 data를 가공하여 사용하도록 수정

sunny-maeng committed:
refactor: gridCell configureCell() 메서드 기능분리

JaeKimdev committed:
refactor: MarketListCell  configureCell() 메서드 기능분리

- generatePriceLabelContent 네이밍 수정
- generateStockLabelContent 네이밍 수정
- Page 데이터 타입 price관련 프로퍼티 Double형으로 수정

sunny-maeng committed:
refactor: Grid, list뷰컨에서 DataSource설정하는 configureDataSource() 메서드에서 스냅샷 적용부분 applySnapshot() 메서드로 기능 분리

sunny-maeng commented:
리댓 외에 추가로, 오토레이아웃 문제 해결했습니다! (https://github.com/yagom-academy/ios-open-market/pull/221/commits/caf6bde9a5c012772885c90b3f185755f2e0d4ed)
> 아 맞다 그리고 앱 실행하다보니 익숙한 문제가 보이더군요 :(
> 고질적으로 맞닥뜨리는 문제죠. 오토레이아웃 문제는 해결 하셨으면 합니다~

스택뷰의 가로 제약이 Content에 걸려있는데 스택뷰 내부에 있는 image의 가로 제약도 Content에제약이 충돌하는 문제였습니다
이미지에 걸린 가로제약을 지워줘서 해결했습니다 :)

JaeKimdev committed:
chore: 코드 컨벤션 통일 및 메서드 호출 순서로 재배치
